src/providers/token-fee-provider.ts

```diff
@@ -0,0 +1,169 @@
+import { ChainId, Token } from '@uniswap/sdk-core';
+import { BigNumber } from '@ethersproject/bignumber';
+import _ from 'lodash';
+
+import { TokenFeeDetector__factory } from '../types/other/factories/TokenFeeDetector__factory';
+import { log, WRAPPED_NATIVE_CURRENCY } from '../util';
+
+import { ICache } from './cache';
+import { IMulticallProvider } from './multicall-provider';
+import { ProviderConfig } from './provider';
+import { DEFAULT_ALLOWLIST } from './token-validator-provider';
+
+export type TokenFeeResult = {
+  buyFeeBps: BigNumber,
+  sellFeeBps: BigNumber,
+}
+
+// on detector failure, assume no fee
+const DEFAULT_TOKEN_FEE_RESULT = {
+  buyFeeBps: BigNumber.from(0),
+  sellFeeBps: BigNumber.from(0),
+};
+
+export interface TokenFeeResults {
+  getFeesByToken(token: Token): TokenFeeResult | undefined;
+}
+
+const FEE_DETECTOR_ADDRESS = '0x57eC54d113719dDE9A90E6bE807524a86560E89D';
+const AMOUNT_TO_FLASH_BORROW = '10000';
+const GAS_LIMIT_PER_VALIDATE = 1_000_000;
+
+/**
+ * Provider for getting token fee data.
+ *
+ * @export
+ * @interface ITokenFeeProvider
+ */
+export interface ITokenFeeProvider {
+  /**
+   * Gets the fees for the token at each address.
+   *
+   * @param tokens The token addresses to fetch fees for.
+   * @param [providerConfig] The provider config.
+   * @returns A token accessor with methods for accessing the tokens.
+   */
+  validateTokens(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenFeeResults>;
+}
+
+export class TokenFeeProvider implements ITokenFeeProvider {
+  private CACHE_KEY = (chainId: ChainId, address: string) =>
+    `token-${chainId}-${address}`;
+
+  private BASE_TOKEN: string;
+
+  constructor(
+    protected chainId: ChainId,
+    protected multicall2Provider: IMulticallProvider,
+    private tokenFeeCache: ICache<TokenFeeResult>,
+    private tokenValidatorAddress = FEE_DETECTOR_ADDRESS,
+    private gasLimitPerCall = GAS_LIMIT_PER_VALIDATE,
+    private amountToFlashBorrow = AMOUNT_TO_FLASH_BORROW,
+    private allowList = DEFAULT_ALLOWLIST
+  ) {
+    this.BASE_TOKEN = WRAPPED_NATIVE_CURRENCY[this.chainId]!.address;
+  }
+
+  public async validateTokens(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenFeeResults> {
+    const tokenAddressToToken = _.keyBy(tokens, 'address');
+    const addressesRaw = _(tokens)
+      .map((token) => token.address)
```

Comment:
nit: would it dedup for the same address but different casing?

Reply:
ah `Token` object actually always eip-55 encodes the address